File: Part 8/NTSA_timefreq_wavelets.m
%%
%     COURSE: Solved problems in neural time series analysis
%    SECTION: Time-frequency analyses
%      VIDEO: Create a family of complex Morlet wavelets
% Instructor: sincxpress.com
%
%%

% time parameters
srate = 1025;
npnts = 2001; % use an odd number!
timev = linspace(-1,1,npnts) * (npnts-1)/srate/2;

% wavelet parameters
minFreq =  2; % Hz
maxFreq = 54; % Hz
numfrex = 99;

% vector of wavelet frequencies
frex = logspace(log10(minFreq),log10(maxFreq),numfrex);

% Gaussian parameters
numCycl = logspace(log10(3),log10(15),numfrex);
fwhms   = logspace(log10(1),log10(.3),numfrex);

%% now create the wavelets

wavefam = zeros(2,numfrex,npnts);

for fi=1:numfrex
    
    % create complex sine wave
    csw = exp( 1i* 2*pi*frex(fi)*timev );
    
    % create the two Gaussians
    s = numCycl(fi) / (2*pi*frex(fi));
    gaus1 = exp( -timev.^2 / (2*s)^2 );
    
    gaus2 = exp( -4*log(2)*timev.^2 / fwhms(fi)^2 );
    
    % now create the complex Morlet wavelets
    wavefam(1,fi,:) = csw .* gaus1;
    wavefam(2,fi,:) = csw .* gaus2;
end

%% image the wavelets

typename = {'numcyc';'FWHM'};

figure(1), clf

for typei=1:2
    subplot(2,3,1+(typei-1)*3)
    contourf(timev,frex,real( squeeze(wavefam(typei,:,:)) ),40,'linecolor','none')
    set(gca,'clim',[-1 1]), axis square
    title([ typename{typei} ': real part' ])
    xlabel('Time (s)'), ylabel('Frequency (Hz)')
    
    subplot(2,3,2+(typei-1)*3)
    contourf(timev,frex,imag( squeeze(wavefam(typei,:,:)) ),40,'linecolor','none')
    set(gca,'clim',[-1 1]), axis square
    title([ typename{typei} ': imag part' ])
    
    subplot(2,3,3+(typei-1)*3)
    contourf(timev,frex,abs( squeeze(wavefam(typei,:,:)) ),40,'linecolor','none')
    set(gca,'clim',[-1 1]), axis square
    title([ typename{typei} ': magnitude' ])
end


%% show an example of one wavelet

figure(2), clf
subplot(211), hold on
plot(timev,squeeze(real(wavefam(1,40,:))),'b')
plot(timev,squeeze(imag(wavefam(1,40,:))),'r')
plot(timev,squeeze(abs(wavefam(1,40,:))),'k')
legend({'real';'imag';'abs'})
title([ 'numcyc wavelet at ' num2str(frex(40)) ' Hz' ])

subplot(212), hold on
plot(timev,squeeze(real(wavefam(2,40,:))),'b')
plot(timev,squeeze(imag(wavefam(2,40,:))),'r')
plot(timev,squeeze(abs(wavefam(2,40,:))),'k')
legend({'real';'imag';'abs'})
title([ 'FWHM wavelet at ' num2str(frex(40)) ' Hz' ])
xlabel('Time (s)')

%% done.

File: Part 8/NTSA_timefreq_wavelets.py
# %%
#     COURSE: Solved problems in neural time series analysis
#    SECTION: Time-frequency analyses
#    TEACHER: Mike X Cohen, sincxpress.com
#      VIDEO: Create a family of complex Morlet wavelets

# !%matplotlib inline
import os
import sys

import matplotlib.pyplot as plt
import numpy as np

# Set figure settings
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from utils import get_cmap, set_fig_dpi, set_style

set_fig_dpi(), set_style()
cmap = get_cmap('parula')
# %%
# Time parameters
srate = 1025
n_pnts = 2001  # Use an odd number! (To have an exact center point)
time = np.linspace(-1, 1, n_pnts) * (n_pnts - 1) / srate / 2

# wavelet parameters
min_freq = 2  # Hz
max_freq = 54  # Hz
n_freqs = 99

# vector of wavelet frequencies
freqs = np.logspace(np.log10(min_freq), np.log10(max_freq), n_freqs)

# Gaussian parameters
n_cycs = np.logspace(np.log10(3), np.log10(15), n_freqs)
fwhms = np.logspace(np.log10(1), np.log10(0.3), n_freqs)

# Now create the wavelets

wave_fam = np.zeros((2, n_freqs, n_pnts), dtype=complex)

for fi in range(n_freqs):
    # Create complex sine wave
    csw = np.exp(1j * 2 * np.pi * freqs[fi] * time)

    # Create the two Gaussians
    s = n_cycs[fi] / (2 * np.pi * freqs[fi])
    gauss1 = np.exp(-(time**2) / (2 * s) ** 2)

    gauss2 = np.exp(-4 * np.log(2) * time**2 / fwhms[fi] ** 2)

    # Now create the complex Morlet wavelets
    wave_fam[0, fi, :] = csw * gauss1
    wave_fam[1, fi, :] = csw * gauss2

# %% Image the wavelets

# Image the wavelets
typename = ['numcyc', 'FWHM']

plt.figure(figsize=(15, 10))

for typei in range(2):
    for part, title_addon in zip(
        [np.real, np.imag, np.abs], ['Real Part', 'Imag Part', 'Magnitude']
    ):
        plt.subplot(2, 3, typei * 3 + (np.real, np.imag, np.abs).index(part) + 1)
        plt.contourf(time, freqs, part(wave_fam[typei, :, :]), 40, cmap=cmap)
        plt.clim([-1, 1])
        plt.title(f'{typename[typei]}: {title_addon}')
        plt.xlabel('Time (s)')
        plt.ylabel('Frequency (Hz)')
        plt.colorbar()

plt.tight_layout()
plt.show()

# Show an example of one wavelet
plt.figure(figsize=(10, 8))

for typei in range(2):
    plt.subplot(2, 1, typei + 1)
    plt.plot(time, np.real(wave_fam[typei, 40, :]), 'b', label='real')
    plt.plot(time, np.imag(wave_fam[typei, 40, :]), 'r', label='imag')
    plt.plot(time, np.abs(wave_fam[typei, 40, :]), 'k', label='abs')
    plt.legend()
    plt.title(f'{typename[typei]} Wavelet at {freqs[40]:.2f} Hz')
    plt.xlabel('Time (s)')

plt.tight_layout()
plt.show()

# %%
